### PR TITLE
Add blog post for Flux3D

### PR DESCRIPTION
**Pull request by nirmal-suthar**
cc: @avik-pal @DhairyaLGandhi

### CODE REVIEW

**Comment by avik-pal** [comment]
@DhairyaLGandhi thoughts?

File: _posts/2020-09-08-Flux3D.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+## Short Walkthrough of the Flux3D.jl
+
+A general Machine learning or Deep learning task usually involves broad steps mentioned below. Therefore this broad steps also act as an effective assessment of where does this package stands in terms of 3D vision tasks.
+
+Let's go through each step briefly!
+
+### 1. Handling 3D DataPoint
+
+Pointcloud, triangle mesh and voxels are three widely used representation for 3D data and most of the 3D vision work is also based on these structures. We can easily initialize Batched structure for PointCloud, TriMesh and VoxelGrid following which we can perform various transforms/metrics and various manipulation.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:35]:
```suggestion
Pointcloud, triangle mesh and voxels are three widely used representation for 3D data and most of the 3D vision work is also based on these structures. To apply transforms to our data, we need to initialise it. 
```

File: _posts/2020-09-08-Flux3D.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        ),
+    )
+
+```
+
+<div style="text-align:center">
+  <img src="/assets/2020-09-08-Flux3D/visualize.png">
+</div>
+
+## Benchmarks for Flux3D.jl
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:165]: I think the benchmarks need to be higher up, with a short sentence talking about how it compares with kaolin.

File: _posts/2020-09-08-Flux3D.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+There are various Julia packages (thanks to awesome Julia community!) which make Flux3D.jl possible. With the help of **Makie** ecosystem, we can easily interact, visualize 3D structures, save the plots and gifs. **NearestNeighbors.jl** which is high performance nearest neighbour search library also makes it possible to perform intense computation metrics like chamfer distance even on CPU. 
+
+## Conclusion
+
+As all the batched structure for PointCloud, TriMesh, and VoxelGrid are stable enough, we can easily define any arbitrary or custom functions on top of it and using Zygote we can easily differentiate through it. And since this package is ready and stable enough, there are a lot of interesting applications of Flux3D.jl along with other interesting packages in Julia like DiffEqFlux, worth exploring.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:186]: "stable enough" is repeated a couple times.

**Review by DhairyaLGandhi — COMMENTED** [review]
Thanks! Sorry for the delay

File: _posts/2020-09-08-Flux3D.md (hunk 2 above)
Comment (nirmal-suthar) [reply _posts/2020-09-08-Flux3D.md:165]: Next paragraph briefly talks about the comparison with the kaolin.

Comment (DhairyaLGandhi) [reply _posts/2020-09-08-Flux3D.md:165]: The message that this approach makes sense is somewhat lost with the real results from the benchmarks being very low down. Maybe just have a sentence here and say that it is discussed further in the post.

Comment (nirmal-suthar) [reply _posts/2020-09-08-Flux3D.md]: made necessary changes

File: _posts/2020-09-08-Flux3D.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
+---
+title: Flux3D.jl, 3D Vision Library in Julia
+author: Nirmal Suthar, Avik Pal, Dhairya Gandhi
+layout: blog
+---
+
+3D computer vision involves various computer vision techniques like classification, segmentation, object detection, super-resolution and many more. Major and important steps in any machine learning or vision-based tasks include designing the model (using Conv and Dense layer), loss objective and training that can always be addressed with sophisticated general-purpose ML ecosystem like FluxML.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:7]:
```suggestion
3D computer vision involves various computer vision techniques like classification, segmentation, object detection, super-resolution and many more. Modeling these problems becomes significantly easier when there is a general purpose differentiation library available, alongside the litany of tools needed to solve a train ML model, which FluxML already supports.
```

File: _posts/2020-09-08-Flux3D.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+Performing 3D vision task also involves preparing datapoint and preparing dataset, applying transforms, and some common 3D metrics for use in designing the loss objective and models, which is usually a boilerplate and can also lead to a bottleneck for performance (in case of inefficient implementation) for a higher number of data points which is usually the case for vision-based tasks.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:9]:
```suggestion
Performing 3D vision tasks involve preparing datasets to fit a certain representation and applying transforms to them. Further, Flux3D exports common 3D metrics for use in designing the loss objective and various 3D layers, which have been optimised for performance for large datasets, typical in vision-based tasks.

File: _posts/2020-09-08-Flux3D.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+Flux3D.jl is a 3D vision library, written completely in Julia. This package utilizes Flux.jl and Zygote.jl as its building blocks for training 3D vision models and for supporting differentiation. This package also has the support of CUDA GPU acceleration with CUDA.jl. Some of the roles of Flux3D are:
+
+* 3D Structures - Accelerated Batched Data structure for PointCloud, VoxelGrid and TriMesh for storing and computation.
+* Transforms -  Multiple transforms for PointCloud, VoxelGrid and TriMesh
+* Metrics - metrics like chamfer_distance, laplacian loss and edge loss for defining loss objectives.
+* Dataset -  ModelNet10/40 (in multiple variants like pointcloud, mesh, voxel) and a Minimal Custom Dataset wrapper.
+* Conversions - Interconversion between different 3D structure (this also help in building different variants of the dataset)
+* Visualize - visualizing PointCloud, TriMesh and VoxelGrid
+* DL model - Implementation some commonly used 3D layers and models.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:25]:
```suggestion
* 3D Deep Learning - Implementations of some commonly used 3D layers and models.
```

File: _posts/2020-09-08-Flux3D.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+## Benchmarks for Flux3D.jl
+
+Kaolin is a popular 3D vision library based on PyTorch. Flux3D.jl is overall faster than Kaolin in terms of applying transforms on PointCloud/TriMesh and comparable with Kaolin in terms of applying metrics. However, there must be work done to improve the back pass in laplacian loss which uses Sparse Arrays. It is evident from the plots of benchmarking that we are easily able to match the performance of Flux3D with Kaolin, and some key-differences between two are also discussed later in this post.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:29]:
```suggestion
Kaolin is a popular 3D vision library based on PyTorch. Flux3D.jl is overall faster than Kaolin in terms of applying transforms on PointCloud/TriMesh and comparable with Kaolin in terms of applying metrics. However, there must be work done to improve the back pass in laplacian loss which uses `SparseArrays`. From the benchmarks, we can tell that Flux3D can outperform Kaolin. Some key-differences between the two are discussed later in this post.
```

File: _posts/2020-09-08-Flux3D.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+julia> dset[1]
+DataPoint:
+    idx: 1
+    data: TriMesh{Float32,UInt32,Array}
+    ground_truth: 1
+    category_name: monitor
+```
+
+We can use conversion transforms to convert the dataset to VoxelGrid. Similarily we can convert VoxelGrid, TriMesh and PointCloud to any 3D structure either using transforms or simply using the regular constructor.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:109]:
```suggestion
We can use conversion transforms to convert the dataset to `VoxelGrid`. Similarily we can convert `VoxelGrid`, `TriMesh` and `PointCloud` to any 3D structure either using transforms or simply using the regular constructor.
```

File: _posts/2020-09-08-Flux3D.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+<p float="middle">
+  <img src="/assets/2020-09-08-Flux3D/bm_pcloud.png">
+  <img src="/assets/2020-09-08-Flux3D/bm_trimesh.png">
+  <img src="/assets/2020-09-08-Flux3D/bm_metrics.png">
+</p>
+
+## Short Walkthrough of the Flux3D.jl
+
+A general Machine learning or Deep learning task usually involves broad steps mentioned below. Therefore this broad steps also act as an effective assessment of where does this package stands in terms of 3D vision tasks.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:39]:
```suggestion
```

File: _posts/2020-09-08-Flux3D.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+Let's go through each step briefly!
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:41]:
```suggestion
A basic training pipeline requires preparing the datasets, defining the model, the loss function and finally training it. 
Let's go through each step along with visualising and evaluating the results.
```

File: _posts/2020-09-08-Flux3D.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    offset: -1
+    Storage type: CuArray
+
+julia> save_trimesh("normalize_aiplane.obj",m2,1)
+
+```
+
+### 2. Preparing Dataset
+
+ModelNet10/40 is one of the widely used 3D dataset, and we can easily access and preprocess this dataset for use in various vision-based tasks using Flux3D. We can also use conversion transforms for using ModelNet10/40 in multiple variants like PointCloud, TriMesh, VoxelGrid.
```
Comment (DhairyaLGandhi) [inline _posts/2020-09-08-Flux3D.md:88]:
```suggestion
ModelNet10/40 is one of the widely used 3D dataset, and we can easily access and preprocess this dataset for use in various vision-based tasks using Flux3D. We can also use conversion transforms for using ModelNet10/40 in multiple representations like `PointCloud`, `TriMesh`, `VoxelGrid`.
```

**Comment by avik-pal** [comment]
Bump

**Comment by DhairyaLGandhi** [comment]
Seems nicer, I think we can go ahead with this for now; can you update the dates and titles for the blog?

**Pull request merged**